go-client/client/tablehandle.go

```diff
@@ -0,0 +1,291 @@
+package client
+
+import (
+	"context"
+	"log"
+
+	"github.com/apache/arrow/go/v8/arrow"
+
+	tablepb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/table"
+	ticketpb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/ticket"
+)
+
+// A reference to a table stored on the deephaven server.
+//
+// It should be eventually released using Release once it is no longer needed.
+type TableHandle struct {
+	client   *Client
+	ticket   *ticketpb2.Ticket
+	schema   *arrow.Schema
+	size     int64
+	isStatic bool
+}
+
+func newTableHandle(client *Client, ticket *ticketpb2.Ticket, schema *arrow.Schema, size int64, isStatic bool) *TableHandle {
+	return &TableHandle{
+		client:   client,
+		ticket:   ticket,
+		schema:   schema,
+		size:     size,
+		isStatic: isStatic,
+	}
+}
+
+// Returns true if this table does not change over time.
+// This will be false for things like streaming tables or timetables.
+func (th *TableHandle) IsStatic() bool {
+	return th.isStatic
+}
+
+// Downloads the current state of the table on the server and returns it as a Record.
+//
+// If a Record is returned successfully, it must be freed later with `record.Release()`
+func (th *TableHandle) Snapshot(ctx context.Context) (arrow.Record, error) {
+	return th.client.snapshotRecord(ctx, th.ticket)
+}
+
+// Creates a new query based on this table. Table operations can be performed on query nodes to create new nodes.
+// A list of query nodes can then be passed to client.ExecQuery() to execute the entire query as a single call.
+func (th *TableHandle) Query() QueryNode {
+	qb := newQueryBuilder(th.client, th)
+	return qb.curRootNode()
+}
+
+// Releases this table handle's resources on the server. The TableHandle is no longer usable after Release is called.
+// Ensure that no other methods (including ones in other goroutines) are using this TableHandle.
+func (th *TableHandle) Release(ctx context.Context) error {
+	if th.client != nil {
+		err := th.client.release(ctx, th.ticket)
+		if err != nil {
+			// This is logged because most of the time this method is used with defer,
+			// which will discard the error value.
+			log.Println("unable to release table:", err.Error())
+			return err
```

Comment:
Errors can be handled in defer, but I think you are correct that most users will not do it.
https://trstringer.com/golang-deferred-function-error-handling/

That raises a question of what we should present as best practices in the examples.